// react-static/dist/commands/start.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var startConfigServer = function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee4() {
    var _this = this;

    var port, configApp;
    return _regenerator2.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            _context4.next = 2;
            return (0, _static.findAvailablePort)(8000);

          case 2:
            port = _context4.sent;

            process.env.STATIC_ENDPOINT = 'http://127.0.0.1:' + port;

            configApp = (0, _express2.default)();


            configApp.use((0, _cors2.default)());

            configApp.get('/siteProps', function () {
              var _ref2 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee(req, res, next) {
                var siteProps;
                return _regenerator2.default.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _context.prev = 0;
                        _context.next = 3;
                        return config.getSiteProps({ dev: true });

                      case 3:
                        siteProps = _context.sent;

                        res.json(siteProps);
                        _context.next = 11;
                        break;

                      case 7:
                        _context.prev = 7;
                        _context.t0 = _context['catch'](0);

                        res.status(500);
                        next(_context.t0);

                      case 11:
                      case 'end':
                        return _context.stop();
                    }
                  }
                }, _callee, _this, [[0, 7]]);
              }));

              return function (_x, _x2, _x3) {
                return _ref2.apply(this, arguments);
              };
            }());

            configApp.get('/getRoutes', function () {
              var _ref3 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee3(req, res, next) {
                var routes;
                return _regenerator2.default.wrap(function _callee3$(_context3) {
                  while (1) {
                    switch (_context3.prev = _context3.next) {
                      case 0:
                        _context3.prev = 0;
                        _context3.next = 3;
                        return config.getRoutes({ dev: true });

                      case 3:
                        routes = _context3.sent;


                        routes.forEach(function (route) {
                          configApp.get('/route' + route.path, function () {
                            var _ref4 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee2(req, res, next) {
                              var initialProps;
                              return _regenerator2.default.wrap(function _callee2$(_context2) {
                                while (1) {
                                  switch (_context2.prev = _context2.next) {
                                    case 0:
                                      _context2.prev = 0;
                                      _context2.next = 3;
                                      return route.getProps({ dev: true });

                                    case 3:
                                      initialProps = _context2.sent;

                                      res.json(initialProps);
                                      _context2.next = 11;
                                      break;

                                    case 7:
                                      _context2.prev = 7;
                                      _context2.t0 = _context2['catch'](0);

                                      res.status(500);
                                      next(_context2.t0);

                                    case 11:
                                    case 'end':
                                      return _context2.stop();
                                  }
                                }
                              }, _callee2, _this, [[0, 7]]);
                            }));

                            return function (_x7, _x8, _x9) {
                              return _ref4.apply(this, arguments);
                            };
                          }());
                        });

                        res.json(routes);
                        _context3.next = 12;
                        break;

                      case 8:
                        _context3.prev = 8;
                        _context3.t0 = _context3['catch'](0);

                        res.status(500);
                        next(_context3.t0);

                      case 12:
                      case 'end':
                        return _context3.stop();
                    }
                  }
                }, _callee3, _this, [[0, 8]]);
              }));

              return function (_x4, _x5, _x6) {
                return _ref3.apply(this, arguments);
              };
            }());

            configApp.listen(port, function (err) {
              if (err) {
                throw err;
              }
            });

          case 9:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this);
  }));

  return function startConfigServer() {
    return _ref.apply(this, arguments);
  };
}();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _webpackDevServer = require('webpack-dev-server');

var _webpackDevServer2 = _interopRequireDefault(_webpackDevServer);

var _formatWebpackMessages = require('react-dev-utils/formatWebpackMessages');

var _formatWebpackMessages2 = _interopRequireDefault(_formatWebpackMessages);

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _server = require('react-dom/server');

var _webpackConfigurator = require('webpack-configurator');

var _webpackConfigurator2 = _interopRequireDefault(_webpackConfigurator);

var _express = require('express');

var _express2 = _interopRequireDefault(_express);

var _cors = require('cors');

var _cors2 = _interopRequireDefault(_cors);

var _DefaultHtml = require('../DefaultHtml');

var _DefaultHtml2 = _interopRequireDefault(_DefaultHtml);

var _copyPublicFolder = require('../copyPublicFolder');

var _copyPublicFolder2 = _interopRequireDefault(_copyPublicFolder);

var _static = require('../static');

var _paths = require('../paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }
//


var first = true;
var compiler = void 0;

var config = (0, _static.getConfig)();

function buildCompiler(_ref5) {
  var port = _ref5.port;

  var webpackConfig = new _webpackConfigurator2.default();

  var webpackConfigDev = require('../webpack.config.dev').default;

  webpackConfig.merge(webpackConfigDev);
  if (config.webpack) {
    config.webpack(webpackConfig, { dev: true });
  }
  var finalWebpackConfig = webpackConfig.resolve();

  compiler = (0, _webpack2.default)(finalWebpackConfig);

  compiler.plugin('invalid', function () {
    console.time(_chalk2.default.green('=> [\u2713] Build Complete'));
    console.log('=> Rebuilding...');
  });

  compiler.plugin('done', function (stats) {
    var messages = (0, _formatWebpackMessages2.default)(stats.toJson({}, true));
    var isSuccessful = !messages.errors.length && !messages.warnings.length;

    if (isSuccessful) {
      console.timeEnd(_chalk2.default.green('=> [\u2713] Build Complete'));
    }

    if (first) {
      first = false;
      console.log(_chalk2.default.green('=> [\u2713] App serving at'), 'http://localhost:' + port);
    }

    if (messages.errors.length) {
      console.log(_chalk2.default.red('Failed to rebuild.'));
      messages.errors.forEach(function (message) {
        console.log(message);
        console.log();
      });
      return;
    }

    if (messages.warnings.length) {
      console.log(_chalk2.default.yellow('Built complete with warnings.'));
      console.log();
      messages.warnings.forEach(function (message) {
        console.log(message);
        console.log();
      });
    }
  });
}

function startDevServer(_ref6) {
  var port = _ref6.port;

  var devServer = new _webpackDevServer2.default(compiler, {
    port: port,
    hot: true,
    disableHostCheck: true,
    contentBase: _paths.DIST,
    publicPath: '/',
    historyApiFallback: true,
    compress: true,
    quiet: true,
    watchOptions: {
      ignored: /node_modules/
    }
  });

  devServer.listen(port, function (err) {
    if (err) {
      return console.log(err);
    }
  });
}

var _default = function () {
  var _ref7 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee5() {
    var _config, port, siteProps, HtmlTemplate, Html, Head, Body, html;

    return _regenerator2.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            _context5.prev = 0;

            process.env.NODE_PATH = _paths.DIST + ':' + _paths.SRC;
            require('module').Module._initPaths();
            _config = (0, _static.getConfig)();
            _context5.next = 6;
            return _fsExtra2.default.remove(_paths.DIST);

          case 6:
            _context5.next = 8;
            return (0, _static.findAvailablePort)(3000);

          case 8:
            port = _context5.sent;
            _context5.next = 11;
            return _config.getSiteProps({ dev: true });

          case 11:
            siteProps = _context5.sent;
            HtmlTemplate = _config.Html || _DefaultHtml2.default;

            Html = function Html(_ref8) {
              var children = _ref8.children,
                  rest = _objectWithoutProperties(_ref8, ['children']);

              return _react2.default.createElement(
                'html',
                _extends({ lang: 'en' }, rest),
                children
              );
            };

            Head = function Head(_ref9) {
              var children = _ref9.children,
                  rest = _objectWithoutProperties(_ref9, ['children']);

              return _react2.default.createElement(
                'head',
                rest,
                children
              );
            };

            Body = function Body(_ref10) {
              var children = _ref10.children,
                  rest = _objectWithoutProperties(_ref10, ['children']);

              return _react2.default.createElement(
                'body',
                rest,
                children,
                _react2.default.createElement('script', { async: true, src: '/app.js' })
              );
            };

            html = (0, _server.renderToStaticMarkup)(_react2.default.createElement(
              HtmlTemplate,
              { staticMeta: {}, Html: Html, Head: Head, Body: Body, siteProps: siteProps },
              _react2.default.createElement('div', { id: 'root' })
            ));
            _context5.next = 19;
            return _fsExtra2.default.outputFile(_path2.default.join(_paths.DIST, 'index.html'), html);

          case 19:

            console.log('=> Copying public directory...');
            console.time(_chalk2.default.green('=> [\u2713] Public directory copied'));
            (0, _copyPublicFolder2.default)(_paths.DIST);
            console.timeEnd(_chalk2.default.green('=> [\u2713] Public directory copied'));

            console.log('=> Building Routes...');
            console.time(_chalk2.default.green('=> [\u2713] Routes Built'));
            _context5.next = 27;
            return _config.getRoutes({ dev: true });

          case 27:
            _config.routes = _context5.sent;
            _context5.next = 30;
            return (0, _static.writeRouteComponentsToFile)(_config.routes);

          case 30:
            console.timeEnd(_chalk2.default.green('=> [\u2713] Routes Built'));

            console.log('=> Building...');
            console.time(_chalk2.default.green('=> [\u2713] Build Complete'));
            _context5.next = 35;
            return startConfigServer();

          case 35:
            buildCompiler({ port: port });
            startDevServer({ port: port });
            _context5.next = 43;
            break;

          case 39:
            _context5.prev = 39;
            _context5.t0 = _context5['catch'](0);

            console.log(_context5.t0);
            process.exit(1);

          case 43:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee5, undefined, [[0, 39]]);
  }));

  return function _default() {
    return _ref7.apply(this, arguments);
  };
}();

exports.default = _default;
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(first, 'first', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(compiler, 'compiler', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(config, 'config', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(startConfigServer, 'startConfigServer', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(buildCompiler, 'buildCompiler', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(startDevServer, 'startDevServer', 'src/commands/start.js');

  __REACT_HOT_LOADER__.register(_default, 'default', 'src/commands/start.js');
}();

;
