// react-static/dist/paths.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PACKAGE = exports.NODE_MODULES = exports.PUBLIC = exports.DIST = exports.SRC = exports.LOCAL_NODE_MODULES = exports.ROOT = undefined;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ROOT = exports.ROOT = _path2.default.resolve(process.cwd());

var resolvePath = function resolvePath(relativePath) {
  return _path2.default.resolve(_path2.default.join(ROOT, relativePath));
};

var LOCAL_NODE_MODULES = exports.LOCAL_NODE_MODULES = _path2.default.resolve(__dirname, '../node_modules');
var SRC = exports.SRC = resolvePath('src');
var DIST = exports.DIST = resolvePath('dist');
var PUBLIC = exports.PUBLIC = resolvePath('public');
var NODE_MODULES = exports.NODE_MODULES = resolvePath('node_modules');
var PACKAGE = exports.PACKAGE = resolvePath('package.json');
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(ROOT, 'ROOT', 'src/paths.js');

  __REACT_HOT_LOADER__.register(resolvePath, 'resolvePath', 'src/paths.js');

  __REACT_HOT_LOADER__.register(LOCAL_NODE_MODULES, 'LOCAL_NODE_MODULES', 'src/paths.js');

  __REACT_HOT_LOADER__.register(SRC, 'SRC', 'src/paths.js');

  __REACT_HOT_LOADER__.register(DIST, 'DIST', 'src/paths.js');

  __REACT_HOT_LOADER__.register(PUBLIC, 'PUBLIC', 'src/paths.js');

  __REACT_HOT_LOADER__.register(NODE_MODULES, 'NODE_MODULES', 'src/paths.js');

  __REACT_HOT_LOADER__.register(PACKAGE, 'PACKAGE', 'src/paths.js');
}();

;
