// react-static/dist/static.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findAvailablePort = exports.writeRouteComponentsToFile = exports.buildXMLandRSS = exports.writeRoutesToStatic = exports.getConfig = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; /* eslint-disable import/no-dynamic-require, react/no-danger */

//


var buildXMLandRSS = exports.buildXMLandRSS = function () {
  var _ref8 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee4(_ref9) {
    var config = _ref9.config;
    var xml, generateXML;
    return _regenerator2.default.wrap(function _callee4$(_context4) {
      while (1) {
        switch (_context4.prev = _context4.next) {
          case 0:
            generateXML = function generateXML(_ref10) {
              var routes = _ref10.routes;

              var xml = '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">';
              routes.forEach(function (route) {
                if (route.noindex) {
                  return;
                }
                xml += '<url>';
                xml += '<loc>' + (route.permalink + '/').replace(/\/{1,}$/gm, '/') + '</loc>';
                xml += route.lastModified ? '<lastmod>' + route.lastModified + '</lastmod>' : '';
                xml += route.priority ? '<priority>' + route.priority + '</priority>' : '';
                xml += '</url>';
              });
              xml += '</urlset>';
              return xml;
            };

            if (config.siteRoot) {
              _context4.next = 4;
              break;
            }

            console.log('\n=> Warning: No \'siteRoot\' defined in \'static.config.js\'!\n=> This is required for both absolute url\'s and a sitemap.xml to be exported.\n');
            return _context4.abrupt('return');

          case 4:
            xml = generateXML({
              routes: config.routes.filter(function (d) {
                return !d.is404;
              }).map(function (route) {
                return _extends({
                  permalink: config.siteRoot + route.path,
                  lastModified: '',
                  priority: 0.5
                }, route);
              })
            });
            _context4.next = 7;
            return _fsExtra2.default.writeFile(_path2.default.join(_paths.DIST, 'sitemap.xml'), xml);

          case 7:
          case 'end':
            return _context4.stop();
        }
      }
    }, _callee4, this);
  }));

  return function buildXMLandRSS(_x4) {
    return _ref8.apply(this, arguments);
  };
}();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _server = require('react-dom/server');

var _fsExtra = require('fs-extra');

var _fsExtra2 = _interopRequireDefault(_fsExtra);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _reactHelmet = require('react-helmet');

var _reactHelmet2 = _interopRequireDefault(_reactHelmet);

var _openport = require('openport');

var _openport2 = _interopRequireDefault(_openport);

var _DefaultHtml = require('./DefaultHtml');

var _DefaultHtml2 = _interopRequireDefault(_DefaultHtml);

var _shared = require('./shared');

var _paths = require('./paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

var defaultEntry = './src/index';
var path404 = '/404';

// Normalize routes with parents, full paths, context, etc.
var normalizeRoutes = function normalizeRoutes(routes) {
  var flatRoutes = [];

  var recurse = function recurse(route) {
    var parent = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : { path: '/' };

    var routePath = route.is404 ? path404 : (0, _shared.pathJoin)(parent.path, route.path);

    if (typeof route.noIndex !== 'undefined') {
      console.log('=> Warning: Route ' + route.path + ' is using \'noIndex\'. Did you mean \'noindex\'?');
      route.noindex = route.noIndex;
    }

    var normalizedRoute = _extends({}, route, {
      path: routePath,
      noindex: typeof route.noindex === 'undefined' ? parent.noindex : route.noindex,
      hasGetProps: !!route.getProps
    });

    if (!normalizedRoute.path) {
      throw new Error('No path defined for route:', normalizedRoute);
    }

    if (route.children) {
      route.children.forEach(function (d) {
        return recurse(d, normalizedRoute);
      });
    }

    delete normalizedRoute.children;

    flatRoutes.push(normalizedRoute);
  };
  routes.forEach(function (d) {
    return recurse(d);
  });

  flatRoutes.forEach(function (route) {
    var found = flatRoutes.find(function (d) {
      return d.path === route.path;
    });
    if (found !== route) {
      console.warn('More than one route is defined for path:', route.path);
    }
  });

  if (!flatRoutes.find(function (d) {
    return d.is404;
  })) {
    flatRoutes.push({
      is404: true,
      path: path404
    });
  }

  return flatRoutes;
};

var getConfig = exports.getConfig = function getConfig() {
  var config = require(_path2.default.resolve(_path2.default.join(process.cwd(), 'static.config.js'))).default;
  return _extends({
    getSiteProps: function getSiteProps() {
      return {};
    }
  }, config, {
    siteRoot: config.siteRoot ? config.siteRoot.replace(/\/{0,}$/g, '') : null,
    getRoutes: function () {
      var _ref = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee() {
        var routes,
            _args = arguments;
        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return config.getRoutes.apply(config, _args);

              case 2:
                routes = _context.sent;
                return _context.abrupt('return', normalizeRoutes(routes));

              case 4:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, undefined);
      }));

      return function getRoutes() {
        return _ref.apply(this, arguments);
      };
    }()
  });
};

var writeRoutesToStatic = function () {
  var _ref2 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee3(_ref3) {
    var config = _ref3.config;
    var userConfig, HtmlTemplate, Comp, siteProps;
    return _regenerator2.default.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            userConfig = getConfig();
            HtmlTemplate = config.Html || _DefaultHtml2.default;
            Comp = require(_path2.default.join(_paths.ROOT, userConfig.entry || defaultEntry)).default;
            _context3.next = 5;
            return userConfig.getSiteProps({ dev: false });

          case 5:
            siteProps = _context3.sent;
            return _context3.abrupt('return', Promise.all(config.routes.map(function () {
              var _ref4 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee2(route) {
                var initialProps, URL, InitialPropsContext, ContextualComp, staticMeta, appHtml, helmet, head, Html, Head, Body, html, htmlFilename, initialPropsFilename, writeHTML, writeJSON;
                return _regenerator2.default.wrap(function _callee2$(_context2) {
                  while (1) {
                    switch (_context2.prev = _context2.next) {
                      case 0:
                        _context2.t0 = route.getProps;

                        if (!_context2.t0) {
                          _context2.next = 5;
                          break;
                        }

                        _context2.next = 4;
                        return route.getProps({ route: route, dev: false });

                      case 4:
                        _context2.t0 = _context2.sent;

                      case 5:
                        initialProps = _context2.t0;

                        if (!initialProps) {
                          _context2.next = 9;
                          break;
                        }

                        _context2.next = 9;
                        return _fsExtra2.default.outputFile(_path2.default.join(_paths.DIST, route.path, 'routeData.json'), JSON.stringify(initialProps || {}));

                      case 9:
                        URL = route.path;

                        // Inject initialProps into static build

                        InitialPropsContext = function (_Component) {
                          _inherits(InitialPropsContext, _Component);

                          function InitialPropsContext() {
                            _classCallCheck(this, InitialPropsContext);

                            return _possibleConstructorReturn(this, (InitialPropsContext.__proto__ || Object.getPrototypeOf(InitialPropsContext)).apply(this, arguments));
                          }

                          _createClass(InitialPropsContext, [{
                            key: 'getChildContext',
                            value: function getChildContext() {
                              return {
                                initialProps: initialProps,
                                siteProps: siteProps,
                                URL: URL
                              };
                            }
                          }, {
                            key: 'render',
                            value: function render() {
                              return this.props.children;
                            }
                          }]);

                          return InitialPropsContext;
                        }(_react.Component);

                        InitialPropsContext.childContextTypes = {
                          initialProps: _propTypes2.default.object,
                          siteProps: _propTypes2.default.object,
                          URL: _propTypes2.default.string
                        };
                        ContextualComp = _react2.default.createElement(
                          InitialPropsContext,
                          null,
                          Comp
                        );
                        staticMeta = {};

                        if (!config.preRenderMeta) {
                          _context2.next = 22;
                          break;
                        }

                        _context2.t1 = _extends;
                        _context2.t2 = {};
                        _context2.t3 = staticMeta;
                        _context2.next = 20;
                        return config.preRenderMeta(ContextualComp);

                      case 20:
                        _context2.t4 = _context2.sent;
                        staticMeta = (0, _context2.t1)(_context2.t2, _context2.t3, _context2.t4);

                      case 22:
                        appHtml = (0, _server.renderToString)(ContextualComp);

                        // Extract head calls using Helmet

                        helmet = _reactHelmet2.default.renderStatic();
                        head = {
                          htmlProps: helmet.htmlAttributes.toComponent(),
                          bodyProps: helmet.bodyAttributes.toComponent(),
                          base: helmet.base.toComponent(),
                          link: helmet.link.toComponent(),
                          meta: helmet.meta.toComponent(),
                          noscript: helmet.noscript.toComponent(),
                          script: helmet.script.toComponent(),
                          style: helmet.style.toComponent(),
                          title: helmet.title.toComponent()
                        };


                        staticMeta.head = head;

                        if (!config.postRenderMeta) {
                          _context2.next = 34;
                          break;
                        }

                        _context2.t5 = _extends;
                        _context2.t6 = {};
                        _context2.t7 = staticMeta;
                        _context2.next = 32;
                        return config.postRenderMeta(appHtml);

                      case 32:
                        _context2.t8 = _context2.sent;
                        staticMeta = (0, _context2.t5)(_context2.t6, _context2.t7, _context2.t8);

                      case 34:
                        Html = function Html(_ref5) {
                          var children = _ref5.children,
                              rest = _objectWithoutProperties(_ref5, ['children']);

                          return _react2.default.createElement(
                            'html',
                            _extends({ lang: 'en' }, head.htmlprops, rest),
                            children
                          );
                        };

                        Head = function Head(_ref6) {
                          var children = _ref6.children,
                              rest = _objectWithoutProperties(_ref6, ['children']);

                          return _react2.default.createElement(
                            'head',
                            rest,
                            head.base,
                            head.link,
                            head.meta,
                            head.noscript,
                            head.script,
                            head.style,
                            head.title,
                            children
                          );
                        };

                        Body = function Body(_ref7) {
                          var children = _ref7.children,
                              rest = _objectWithoutProperties(_ref7, ['children']);

                          return _react2.default.createElement(
                            'body',
                            _extends({}, head.bodyProps, rest),
                            children,
                            _react2.default.createElement('script', {
                              type: 'text/javascript',
                              dangerouslySetInnerHTML: {
                                __html: 'window.__routeData = ' + JSON.stringify({
                                  path: route.path,
                                  initialProps: initialProps,
                                  siteProps: siteProps
                                })
                              }
                            }),
                            _react2.default.createElement('script', { async: true, src: '/app.js' })
                          );
                        };

                        html = '<!DOCTYPE html>' + (0, _server.renderToString)(_react2.default.createElement(
                          HtmlTemplate,
                          {
                            staticMeta: staticMeta,
                            Html: Html,
                            Head: Head,
                            Body: Body,
                            siteProps: siteProps
                          },
                          _react2.default.createElement('div', { id: 'root', dangerouslySetInnerHTML: { __html: appHtml } })
                        ));


                        if (config.siteRoot) {
                          html = html.replace(/(href=["'])(\/[^/])/gm, '$1' + config.siteRoot + '$2');
                        }

                        htmlFilename = route.is404 ? _path2.default.join(_paths.DIST, '404.html') : _path2.default.join(_paths.DIST, route.path, 'index.html');
                        initialPropsFilename = _path2.default.join(_paths.DIST, route.path, 'routeData.json');
                        writeHTML = _fsExtra2.default.outputFile(htmlFilename, html);
                        writeJSON = _fsExtra2.default.outputFile(initialPropsFilename, JSON.stringify({
                          initialProps: initialProps
                        }));
                        _context2.next = 45;
                        return Promise.all([writeHTML, writeJSON]);

                      case 45:
                      case 'end':
                        return _context2.stop();
                    }
                  }
                }, _callee2, undefined);
              }));

              return function (_x3) {
                return _ref4.apply(this, arguments);
              };
            }())));

          case 7:
          case 'end':
            return _context3.stop();
        }
      }
    }, _callee3, undefined);
  }));

  return function writeRoutesToStatic(_x2) {
    return _ref2.apply(this, arguments);
  };
}();

exports.writeRoutesToStatic = writeRoutesToStatic;
var writeRouteComponentsToFile = exports.writeRouteComponentsToFile = function () {
  var _ref11 = _asyncToGenerator( /*#__PURE__*/_regenerator2.default.mark(function _callee5(routes) {
    var templates, standardRoutes, notFoundRoute, file, filepath, now, then;
    return _regenerator2.default.wrap(function _callee5$(_context5) {
      while (1) {
        switch (_context5.prev = _context5.next) {
          case 0:
            templates = [];

            routes = routes.filter(function (d) {
              return d.component;
            });
            routes.forEach(function (route) {
              if (!templates.includes(route.component)) {
                templates.push(route.component);
              }
            });

            standardRoutes = routes.filter(function (d) {
              return !d.is404;
            });
            notFoundRoute = routes.find(function (d) {
              return d.is404;
            });
            file = '\n    import React, { Component } from \'react\'\n    import { Switch, Route } from \'react-router-dom\'\n\n    ' + templates.map(function (template) {
              return 'import ' + template.replace(/[^a-zA-Z]/g, '_') + ' from \'../' + template + '\'';
            }).join('\n') + '\n\n    export default class Routes extends Component {\n      render () {\n        return (\n          <Switch>\n              ' + standardRoutes.map(function (route) {
              return '<Route exact path={\'' + route.path + '\'} component={' + route.component.replace(/[^a-zA-Z]/g, '_') + '} />';
            }).join(',\n') + '\n              ' + (notFoundRoute ? '<Route component={' + notFoundRoute.component.replace(/[^a-zA-Z]/g, '_') + '} />' : '') + '\n          </Switch>\n        )\n      }\n    }\n  ';
            filepath = _path2.default.resolve(_paths.DIST, 'react-static-routes.js');
            _context5.next = 9;
            return _fsExtra2.default.remove(filepath);

          case 9:
            _context5.next = 11;
            return _fsExtra2.default.writeFile(filepath, file);

          case 11:
            now = Date.now() / 1000;
            then = now - 1000;

            _fsExtra2.default.utimesSync(filepath, then, then);

          case 14:
          case 'end':
            return _context5.stop();
        }
      }
    }, _callee5, undefined);
  }));

  return function writeRouteComponentsToFile(_x5) {
    return _ref11.apply(this, arguments);
  };
}();

var findAvailablePort = exports.findAvailablePort = function findAvailablePort(start) {
  return new Promise(function (resolve, reject) {
    return _openport2.default.find({
      startingPort: start,
      endingPort: start + 1000
    }, function (err, port) {
      if (err) {
        return reject(err);
      }
      resolve(port);
    });
  });
};
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(defaultEntry, 'defaultEntry', 'src/static.js');

  __REACT_HOT_LOADER__.register(path404, 'path404', 'src/static.js');

  __REACT_HOT_LOADER__.register(normalizeRoutes, 'normalizeRoutes', 'src/static.js');

  __REACT_HOT_LOADER__.register(getConfig, 'getConfig', 'src/static.js');

  __REACT_HOT_LOADER__.register(writeRoutesToStatic, 'writeRoutesToStatic', 'src/static.js');

  __REACT_HOT_LOADER__.register(buildXMLandRSS, 'buildXMLandRSS', 'src/static.js');

  __REACT_HOT_LOADER__.register(writeRouteComponentsToFile, 'writeRouteComponentsToFile', 'src/static.js');

  __REACT_HOT_LOADER__.register(findAvailablePort, 'findAvailablePort', 'src/static.js');
}();

;
