// react-static/dist/webpack.config.prod.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

//


var _webpack = require('webpack');

var _webpack2 = _interopRequireDefault(_webpack);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _caseSensitivePathsWebpackPlugin = require('case-sensitive-paths-webpack-plugin');

var _caseSensitivePathsWebpackPlugin2 = _interopRequireDefault(_caseSensitivePathsWebpackPlugin);

var _webpackBundleAnalyzer = require('webpack-bundle-analyzer');

var _rules = require('./rules');

var _rules2 = _interopRequireDefault(_rules);

var _static = require('./static');

var _paths = require('./paths');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var defaultEntry = './src/index';
var config = (0, _static.getConfig)();

var _default = {
  context: _path2.default.resolve(__dirname, '../node_modules'),
  entry: [_path2.default.resolve(_paths.ROOT, config.entry || defaultEntry)],
  output: {
    filename: 'app.js',
    path: _paths.DIST,
    publicPath: '/'
  },
  module: {
    rules: _rules2.default
  },
  resolve: {
    modules: [_paths.NODE_MODULES, _path2.default.resolve(__dirname, '../node_modules'), _paths.SRC, _paths.DIST],
    extensions: ['.js', '.json', '.jsx']
    // alias: {
    //   'react-static-routes': path.resolve(DIST, 'react-static-routes.js'),
    // },
  },
  plugins: [new _webpack2.default.EnvironmentPlugin(_extends({}, process.env, {
    NODE_ENV: 'production'
  })), new _caseSensitivePathsWebpackPlugin2.default(), new _webpack2.default.optimize.UglifyJsPlugin(), config.bundleAnalyzer ? new _webpackBundleAnalyzer.BundleAnalyzerPlugin() : null].filter(function (d) {
    return d;
  }),

  devtool: 'source-map'
};
exports.default = _default;
;

var _temp = function () {
  if (typeof __REACT_HOT_LOADER__ === 'undefined') {
    return;
  }

  __REACT_HOT_LOADER__.register(defaultEntry, 'defaultEntry', 'src/webpack.config.prod.js');

  __REACT_HOT_LOADER__.register(config, 'config', 'src/webpack.config.prod.js');

  __REACT_HOT_LOADER__.register(_default, 'default', 'src/webpack.config.prod.js');
}();

;
